### DISCUSSION TITLE
Directus Storage Adapter AWS S3 error

### DISCUSSION THREAD

**Discussion by Juanjo-GEx** [directus/directus#15566] [Archive]
### Describe the Bug

We have used the included adapter for AWS-S3 to store our files. If we save a record manually we observe that it is saved correctly.
[Manual storage](https://www.loom.com/share/af7f83708a0540a59ba1fba4778bd746)

The problem occurs when the request is made via the Rest API and the file has been generated automatically.
[Automatic storage](https://www.loom.com/share/4c777e8ac3f5484c971c5db62152f605)

### To Reproduce

A hook of type _action_ has been created. When a _create event_ is detected in the _urls_ collection, the hook generates a **short url** and an **empty file** with the **redirection metadata**. 

```javascript
const createFileObject = () => {
    const fileObject = {					
	    "storage": "s3",
	    "filename_disk": `${shortURL}.txt`,
	    "filename_download": `${shortURL}.txt`,
	    "title": shortURL,
	    "type": "text/plain",
	    "uploaded_by": accountability.user,
	    "filesize": 0,
	    "metadata": {
		    "x-amz-website-redirect-location": `${endsToSlash(input.payload.domain)}${shortURL}`
	    }
    }
    return fileObject
}
```

```javascript
await recordService.updateOne(input.key, {
	slug: shortURL,
	redirection: fileObject
});
```

It is observed that although the hook works correctly and saves the record in its table, the created file does not appear in the amazon s3 bucket.

### Errors Shown

The automatically generated file is not saved in amazon s3.

### What version of Directus are you using?

9.17.4

### What version of Node.js are you using?

16.16.0

### What database are you using?

SQLite 5.0.11

### What browser are you using?

Chrome

### How are you deploying Directus?

locally

**Comment by azrikahar** [comment]
I believe there is a confusion here, as the "automatically generated file" as mentioned isn't really generated, at least based on my current understanding of the setup here.

When you manually save a record and upload the file via the App as seen here:

![chrome_5h3jyMzs9p](https://user-images.githubusercontent.com/42867097/190079627-25f8e2e6-93c2-4e08-83ae-a16ae807e105.png)

You will make a POST request to `/files` which passes through the `multipartHandler` request handler:

https://github.com/directus/directus/blob/9b24233aa4770024bb7a853142e596d00c3fe654/api/src/controllers/files.ts#L120-L123

which then actually calls `uploadOne` with the file stream which performs the file upload:

https://github.com/directus/directus/blob/9b24233aa4770024bb7a853142e596d00c3fe654/api/src/controllers/files.ts#L90

However in your "automatically generated file" setup, you are only sending the json object which contains the _metadata_ of the file, but _not_ the file itself.


```js
{					
	    "storage": "s3",
	    "filename_disk": `${shortURL}.txt`,
	    "filename_download": `${shortURL}.txt`,
	    "title": shortURL,
	    "type": "text/plain",
	    "uploaded_by": accountability.user,
	    "filesize": 0,
	    "metadata": {
		    "x-amz-website-redirect-location": `${endsToSlash(input.payload.domain)}${shortURL}`
	    }
}
```

(there is no actual file being created/streamed above)

---

In order to actually perform the file upload, you will need to either:

- create the file (for example using tmp: https://www.npmjs.com/package/tmp)
- read the file somewhere

to get the file stream, then create a `FilesService` service and use it to call `.uploadOne()` with aforementioned file stream with the necessary payload (which is what you have in `createFileObject()`).

Hope that clears things up 👍

**Comment by Juanjo-GEx** [comment]
The explanation is perfect mate!
The problem I was having is perfectly understandable. The bad thing is that I have an error when saving the file.

```javascript
const fileTmp = tmp.fileSync();

const fileUpload = await filesService.uploadOne(fileTmp, {
					storage: shortURL,
					filename_disk: `${shortURL}.txt`,
					filename_download: `${shortURL}.txt`,
					title: shortURL,
					type: "text/plain",
					uploaded_by: accountability.user,
					filesize: 0,
					metadata: {
						x-amz-website-redirect-location: `${endsToSlash(input.payload.domain)}${shortURL}`
					}
});
console.log(fileUpload)

await recordService.updateOne(input.key, {
	slug: shortURL,
	redirection: fileUpload
});
```

ERROR
```console
Couldn't save file 7b9b2787-5361-4101-9fcc-a5d5fdf15ead.txt
E_INVALID_CONFIG: Make sure to define config for [object Promise] disk
An error was thrown while executing action "urls.items.create"
Error: Couldn't save file 7b9b2787-5361-4101-9fcc-a5d5fdf15ead.txt
```

**Reply by azrikahar** [reply]
Glad the explanation is helpful! In hindsight, I should have probably bring up `tmp-promise` instead of `tmp`, since we're already using `tmp-promise` in the API. (and `tmp.fileSync()` will block the main thread if I'm not mistaken)

Perhaps you can try referring to the current implementation of how the API does the async export file to library logic then adapt it to your need, since it should be mostly the same as what you're aiming to achieve.

Here are some relevant lines of code:

1. Using `tmp-promise` instead of `tmp`:
    https://github.com/directus/directus/blob/9b24233aa4770024bb7a853142e596d00c3fe654/api/src/services/import-export.ts#L13
    
2. Create the tmp file, then destructure `path` and `cleanup` from it:

    https://github.com/directus/directus/blob/9b24233aa4770024bb7a853142e596d00c3fe654/api/src/services/import-export.ts#L194
    
3. This will be the bulk of what you were already doing so you can refer and modify accordingly, but an important note is that we're using `createReadStream(path)` to pass the file stream of the tmp file (which might be the potential issue in your previous implementation):


    https://github.com/directus/directus/blob/9b24233aa4770024bb7a853142e596d00c3fe654/api/src/services/import-export.ts#L246-L264
    
4. Then make sure to clean up that tmp file to prevent your local disk from filling up with tons of tmp files (`cleanup` is the function from step 2):

    https://github.com/directus/directus/blob/9b24233aa4770024bb7a853142e596d00c3fe654/api/src/services/import-export.ts#L281

**Comment by Juanjo-GEx** [comment]
Woww you are really good.
**It keeps giving the same error,** although he file is generated correctly. The payload fields have also been added successfully. This is the console output of `createReadStream(path)`: 

```console
ReadStream {
  fd: null,
  path: 'C:\\Users\\jalonso\\AppData\\Local\\Temp\\tmp-13652-A8k34WT0np4h',
  flags: 'r',
  mode: 438,
  start: undefined,
  end: Infinity,
  pos: undefined,
  bytesRead: 0,
  closed: false,
  _readableState: ReadableState {
    objectMode: false,
    highWaterMark: 65536,
    buffer: BufferList { head: null, tail: null, length: 0 },
    length: 0,
    pipes: [],
    flowing: null,
    ended: false,
    endEmitted: false,
    reading: false,
    constructed: false,
    sync: true,
    needReadable: false,
    emittedReadable: false,
    readableListening: false,
    resumeScheduled: false,
    errorEmitted: false,
    emitClose: true,
    autoDestroy: true,
    destroyed: false,
    errored: null,
    closed: false,
    closeEmitted: false,
    defaultEncoding: 'utf8',
    awaitDrainWriters: null,
    multiAwaitDrain: false,
    readingMore: false,
    dataEmitted: false,
    decoder: null,
    encoding: null,
    [Symbol(kPaused)]: null
  },
  _events: [Object: null prototype] {
    [Symbol(kConstruct)]: [Function: bound onceWrapper] { listener: [Function (anonymous)] }
  },
  _eventsCount: 1,
  _maxListeners: undefined,
  [Symbol(kFs)]: {
    appendFile: [Function: appendFile],
    appendFileSync: [Function: appendFileSync],
    access: [Function: access],
    accessSync: [Function: accessSync],
    chown: [Function: chown],
    chownSync: [Function: chownSync],
    chmod: [Function: chmod],
    chmodSync: [Function: chmodSync],
    close: [Function: close],
    closeSync: [Function: closeSync],
    copyFile: [Function: copyFile],
    copyFileSync: [Function: copyFileSync],
    cp: [Function: cp],
    cpSync: [Function: cpSync],
    createReadStream: [Function: createReadStream],
    createWriteStream: [Function: createWriteStream],
    exists: [Function: exists],
    existsSync: [Function: existsSync],
    fchown: [Function: fchown],
    fchownSync: [Function: fchownSync],
    fchmod: [Function: fchmod],
    fchmodSync: [Function: fchmodSync],
    fdatasync: [Function: fdatasync],
    fdatasyncSync: [Function: fdatasyncSync],
    fstat: [Function: fstat],
    fstatSync: [Function: fstatSync],
    fsync: [Function: fsync],
    fsyncSync: [Function: fsyncSync],
    ftruncate: [Function: ftruncate],
    ftruncateSync: [Function: ftruncateSync],
    futimes: [Function: futimes],
    futimesSync: [Function: futimesSync],
    lchown: [Function: lchown],
    lchownSync: [Function: lchownSync],
    lchmod: undefined,
    lchmodSync: undefined,
    link: [Function: link],
    linkSync: [Function: linkSync],
    lstat: [Function: lstat],
    lstatSync: [Function: lstatSync],
    lutimes: [Function: lutimes],
    lutimesSync: [Function: lutimesSync],
    mkdir: [Function: mkdir],
    mkdirSync: [Function: mkdirSync],
    mkdtemp: [Function: mkdtemp],
    mkdtempSync: [Function: mkdtempSync],
    open: [Function: open],
    openSync: [Function: openSync],
    opendir: [Function: opendir],
    opendirSync: [Function: opendirSync],
    readdir: [Function: readdir],
    readdirSync: [Function: readdirSync],
    read: [Function: read],
    readSync: [Function: readSync],
    readv: [Function: readv],
    readvSync: [Function: readvSync],
    readFile: [Function: readFile],
    readFileSync: [Function: readFileSync],
    readlink: [Function: readlink],
    readlinkSync: [Function: readlinkSync],
    realpath: [Function: realpath] { native: [Function (anonymous)] },
    realpathSync: [Function: realpathSync] { native: [Function (anonymous)] },
    rename: [Function: rename],
    renameSync: [Function: renameSync],
    rm: [Function: rm],
    rmSync: [Function: rmSync],
    rmdir: [Function: rmdir],
    rmdirSync: [Function: rmdirSync],
    stat: [Function: stat],
    statSync: [Function: statSync],
    symlink: [Function: symlink],
    symlinkSync: [Function: symlinkSync],
    truncate: [Function: truncate],
    truncateSync: [Function: truncateSync],
    unwatchFile: [Function: unwatchFile],
    unlink: [Function: unlink],
    unlinkSync: [Function: unlinkSync],
    utimes: [Function: utimes],
    utimesSync: [Function: utimesSync],
    watch: [Function: watch],
    watchFile: [Function: watchFile],
    writeFile: [Function: writeFile],
    writeFileSync: [Function: writeFileSync],
    write: [Function: write],
    writeSync: [Function: writeSync],
    writev: [Function: writev],
    writevSync: [Function: writevSync],
    Dir: [class Dir],
    Dirent: [class Dirent],
    Stats: [Function: Stats],
    ReadStream: [Getter/Setter],
    WriteStream: [Getter/Setter],
    FileReadStream: [Getter/Setter],
    FileWriteStream: [Getter/Setter],
    _toUnixTimestamp: [Function: toUnixTimestamp],
    F_OK: 0,
    R_OK: 4,
    W_OK: 2,
    X_OK: 1,
    constants: [Object: null prototype] {
      UV_FS_SYMLINK_DIR: 1,
      UV_FS_SYMLINK_JUNCTION: 2,
      O_RDONLY: 0,
      O_WRONLY: 1,
      O_RDWR: 2,
      UV_DIRENT_UNKNOWN: 0,
      UV_DIRENT_FILE: 1,
      UV_DIRENT_DIR: 2,
      UV_DIRENT_LINK: 3,
      UV_DIRENT_FIFO: 4,
      UV_DIRENT_SOCKET: 5,
      UV_DIRENT_CHAR: 6,
      UV_DIRENT_BLOCK: 7,
      S_IFMT: 61440,
      S_IFREG: 32768,
      S_IFDIR: 16384,
      S_IFCHR: 8192,
      S_IFLNK: 40960,
      O_CREAT: 256,
      O_EXCL: 1024,
      UV_FS_O_FILEMAP: 536870912,
      O_TRUNC: 512,
      O_APPEND: 8,
      F_OK: 0,
      R_OK: 4,
      W_OK: 2,
      X_OK: 1,
      UV_FS_COPYFILE_EXCL: 1,
      COPYFILE_EXCL: 1,
      UV_FS_COPYFILE_FICLONE: 2,
      COPYFILE_FICLONE: 2,
      UV_FS_COPYFILE_FICLONE_FORCE: 4,
      COPYFILE_FICLONE_FORCE: 4
    },
    promises: [Getter]
  },
  [Symbol(kIsPerformingIO)]: false,
  [Symbol(kCapture)]: false
}
```

**Reply by azrikahar** [reply]
Hmm I not sure can I spot anything from this stream output, but just to clarify, everything works except the error `E_INVALID_CONFIG: Make sure to define config for [object Promise] disk` that you shared earlier?

If that is the case, the `[object Promise]` stands out to me. I assume you are passing a promise by mistake instead of a string? This is where the error originates from: 

https://github.com/directus/directus/blob/9b24233aa4770024bb7a853142e596d00c3fe654/packages/drive/src/exceptions/InvalidConfig.ts#L12-L14

I think instead of doing `storage: shortURL`, you should do `storage: "s3"`. `storage` should be a storage location (what you configured in `STORAGE_LOCATIONS`), not a URL.

**Comment by Juanjo-GEx** [comment]
Yes yes! That was the mistake. I don't know how that `shortURL` could get there if I always had it right config the `storage: "s3"`.
**Everything works correctly now.**
Great work and help Rik. Thanks a lot!!
